### ISSUE TITLE
Merge fork-pavelvasev

### ISSUE THREAD

**Issue by ChALkeR** [qmlweb/qmlweb#32]
Merge fork by @pavelvasev.
That is actually based on #19, I excluded the commits that are already present in #19.

- [x] c567f68 README.md: Fix color name in tutorial — **Rejected**, Readme-only change, inapplicable.
- [x] **[listmodel]** 04bd0b1 ListModel: Notify dependent properties about list element changes with .set(). — **Merged** in #206.
- [x] **[upflow]** 40895fb upflow option — **Discarded** by @pavelvasev.
- [x] **[upflow]** 88839ab did it! up flow contexts in qmlweb! — **Discarded** by @pavelvasev.
- [x] **[context]** a623154 context comments — **Discarded** by @pavelvasev. Comments only.
- [x] **[aliaschangeback, textarea, textinput, button]** dce7d96 fixes from viewlang — Button changes are **discarded** by @pavelvasev, TextInput  signal changes **rejected** (not compatible with Qt QML). TextEdit padding fixes — **Merged** in #208. Alias fixes — **Merged** in #195.
- [x] **[checkbox]** 6bbfa83 checkbox fixes — **Merged** in #198.
- [x] **[video]** 816f334 fix crop mode for video — **Rejected**, can't reproduce any issue without it, @pavelvasev doesn't remember what that fixes. Perhaps that is controls-related, which we don't need.
- [x] **[recursiveprops]** 595a29f recursive mode for $initializePropertyBindings — **Merged** in #233.
- [x] **[createqmlobject]** a5efa1f createQmlObject fix — **Merged** in #171.
- [x] **[context, signals]** cfb28a0 discard last changes and refactor completed signals — partially **already merged**, partially **merged** in #171.
- [x] **[outercontext]** 561b202 propagate properties of component root objects to outer context — **Postponed** in #241.
- [x] **[context]** 871fbf0 Component context optimization.. maybe we do not need it? — **Merged** in #249.
- [x] **[context]** bbcf79f some bugfixes on context — **Merged** in #212.
- [x] **[fontbug]** 6934ddd small bugfix — **Rejected**, already done in b0fdbe20ad06e3424d5ef34806f517af1e19a2ad (#186).
- [x] **[optimize]** 690c82b small optimization — **Rejected**, is not worth it at the current state, optimizations would be performed at a later stage.
- [x] **[debug]** bae9d88 keep file path information in exceptions during qml parse — **Merged** in #205.
- [x] **[import]** f869b53 qmldirsContents contains QtQuick by default — **Merged** in #68.
- [x] **[binding]** 57c847f binding scope bugfix — **Discarded** by @pavelvasev.
- [x] **[import]** 25c4cf4 blob path fix — **Merged** in #68.
- [x] **[video]** 13af286 Video enhancements — **Rejected**, those are non-standard.
- [x] **[video]** 1ac2f5a Video control — **Already present**, this is just an import of @Plaristote Video implementation without any actual differences.
- [x] **[createqmlobject, import, #qobjectdelete]** 4c3bcc6 createQmlObject and fixes — **Merged** in be11b634, a081bf74, #171, #197, #196.
- [x] **[comment, import]** 22be36a small changes — **merged** in fdf6f97f, #169.
- [x] **[import]** b91417c small bugfix — **Merged** in #68.
- [x] **[recursiveprops]** 0aab767 Bugfixes on recursive properties — **Merged** in #233.
- [x] **[recursiveprops]** 09b8243 Fixes to recursive properties dependency checking — **Merged** in #233.
- [x] **[recursiveprops]** 0e80a6e Recursive property calculation on init — **Merged** in #233.
- [x] **[import]** 9639deb Fix loading qmldir file when current dir is empty — **Merged** in #68.
- [x] **[import]** 7302347 2 fixes — **Merged** in #68.
- [x] **[githubimport]** 8cea5b6 fix — **Rejected** PR in #194, see #203 instead.
- [x] **[githubimport]** cc54d05 fixes to import code from github — **Rejected** PR in #194, see #203 instead.
- [x] **[checkbox]** 04746da Checkbox bugfix — **Merged** in #125.
- [x] **[repeater]** 497fbbd temp bugfix — **Merged** in #96.
- [x] **[aliassignals]** 780450b Aliases now emit changes signals — **Merged** in #195.
- [x] **[import]** e866988 Big bugfix — **Merged** in #68.
- [x] **[import]** 031c7c9 Uglybug in resolvedUrl fixed — **Merged** in #68.
- [x] **[textinput]** 659f381 Uglybug fix of text input — **Merged** in #124.
- [x] **[color]** 43c22d5 Fix QMLColor function and Rectangle color handling — **Merged** in #123.
- [x] **[compdestr]** 91eb23c Now $delete also set's parent to undefined — **Merged** in #115.
- [x] **[repeater]** 9dd3d75 Repeater bugfix — **Merged** in #96.
- [x] **[compdestr]** 04afa40 Component.onDestruction — **Merged** in #115.
- [x] **[repeater]** a743f6d Bugfix for Repeater — **Merged** in #96.
- [x] **[import]** 22ce160 Bugfix — **Merged** in #68.
- [x] **[import]** 281bf5b Qt.resolvedUrl function — **Merged** in #68.
- [x] **[import]** ac7ad8c Create engine.qmldirs and engine.qmldirsContent vars earlier — **Merged** in #68.
- [x] **[domtag]** 54de7de Allow to specify html tag name for component — **Rejected**, see #178. This functionality should be moved to a special element, e.g. in QmlWeb.Dom — see #182.
- [x] **[import]** 48ace8c Import statemement test and small bugfix — **Merged** in #68.
- [x] **[import]** 4470475 addModulePath method and some fixes — **Merged** in #68.
- [x] **[import]** 409b7b5 Initial import implementation — **Merged** in #68.
- [x] **[import]** adea24c Keep info that import statement uses qualified dotted notation — **Merged** in #68.
- [x] **[import]** 5459537 Comment on how import statements work — **Merged** in #68.
- [x] **[import]** c65dae1 Bugfix in readQmlDir — **Merged** in #68.
- [x] **[sizing]** a339ee7 Root element resizing fixes — **Merged** in #204.
- [x] **[import]** 3deb3bb Allow Qt.createComponent to load from urls too — **Merged** in #68.
- [x] **[import]** 7ec1e3b Revert to old exception-style behavior in getUrlContents — **Merged** in #68.
- [x] e89b83e Merge remote-tracking branch 'upstream/master' — **Rejected**, this is just 08a45584e37b4493f4c15546964b28df9e9d5334, which is present in #19.
- [x] **[import]** 2131576 bugfix for engine::addImportPath — **Merged** in #68.
- [x] **[import]** d3b575c engine::setImportPathList method — **Merged** in #68.
- [x] **[import]** bc1ba46 Less error logging for Qt.createComponent — **Merged** in #68.
- [x] **[import]** 7475fce addImportPath etc — **Merged** in #68.
- [x] **[import]** 72b7834 getUrlContents changed to return "" on failed gets. — **Merged** in #68.

**Comment by ChALkeR** [comment]
This actually needs some squashing

**Comment by henrikrudstrom** [comment]
@pavelvasev, could you tell a bit more about your branch
especially curious about 88839abba7479b6f5d31fae9dc94bbd648cbe382 
>did it!  up flow contexts in qmlweb!

**Comment by pavelvasev** [comment]
Hi @henrikrudstrom please note 40895fba55f99ddb5f96ef13dfd4449c86871b11 together with 88839abba7479b6f5d31fae9dc94bbd648cbe382 

"Upflow" is a feature of QML when properties from base components are accessible directly in ancestors components's sub-objects. Sorry for ugly explanation, however 
there is a test case for that: https://github.com/pavelvasev/qmlweb/tree/master/test/testpad/UpFlow

This test case passes with qmlscene tool, and didn't pass with qmlweb.

Probably the same for aliases, methods, signals - I haven't checked.

But later, I decided than that this feature is not mandatory, that's why there is an option. At least all my code works well without the feature after small refactoring. 

The feature degrades performance, I think, because introduces additional context lookup (via with). On the other hand, it gives more compatibility with original QML behaviour.

**Comment by ChALkeR** [comment]
> The feature degrades performance, I think, because introduces additional context lookup (via with).

I don't think that we should worry about performance too much for now, that could be optimized later, after the merges. There are much work to do there.

Le'ts focus on merges and compatibility first ;-).

**Comment by pavelvasev** [comment]
@ChALkeR ok, but let it be the option, turned off by default ;-)

**Comment by pavelvasev** [comment]
Here are commits tagged by logical topics. Every topic has a description (below in comment). Looks like every topic might be pushed as separate PR.

## Commits

- [ ] 04bd0b1 ListModel: Notify dependent properties about list element changes with .set() **@listmodel**.
- [ ] 40895fb upflow option **@upflow**
- [ ] 88839ab did it! up flow contexts in qmlweb! **@upflow**
- [ ] a623154 context comments **@context**
- [ ] dce7d96 fixes from viewlang **@bindingloop** **@textarea** **@textinput** **@button**
- [ ] 6bbfa83 checkbox fixes **@checkbox** 
- [ ] 816f334 fix crop mode for video **@video**
- [ ] 595a29f recursive mode for $initializePropertyBindings **@recursiveprops**
- [ ] a5efa1f createQmlObject fix **@createqmlobject**
- [ ] cfb28a0 discard last changes and refactor completed signals **@context** **@signals**
- [ ] 561b202 propagate properties of component root objects to outer context **@outercontext**
- [ ] 871fbf0 Component context optimization.. maybe we do not need it? **@context**
- [ ] bbcf79f some bugfixes on context **@context**
- [ ] 6934ddd small bugfix **@fontbug**
- [ ] 690c82b small optimization **@optimize**
- [ ] bae9d88 keep file path information in exceptions during qml parse **@debug**
- [ ] f869b53 qmldirsContents contains QtQuick by default **@improt**
- [ ] 57c847f binding scope bugfix **@binding**
- [ ] 25c4cf4 blob path fix **@import**
- [ ] 13af286 Video enhancements **@video**
- [ ] 1ac2f5a Video control **@video**
- [ ] 4c3bcc6 createQmlObject and fixes **@createqmlobject** **@import** **@qobjectdelete**
- [ ] 22be36a small changes **@comment** **@import**
- [ ] b91417c small bugfix **@import**
- [ ] 0aab767 Bugfixes on recursive properties **@recursiveprops**
- [ ] 09b8243 Fixes to recursive properties dependency checking **@recursiveprops**
- [ ] 0e80a6e Recursive property calculation on init **@recursiveprops**
- [ ] 9639deb Fix loading qmldir file when current dir is empty **@import**
- [ ] 7302347 2 fixes **@import**
- [ ] 8cea5b6 fix **@githubimport**
- [ ] cc54d05 fixes to import code from github **@githubimport**
- [ ] 04746da Checkbox bugfix **@checkbox**
- [ ] 497fbbd temp bugfix **@repeater**
- [ ] 780450b Aliases now emit changes signals **@aliassignals**
- [ ] e866988 Big bugfix **@import**
- [ ] 031c7c9 Uglybug in resolvedUrl fixed **@import**
- [ ] 659f381 Uglybug fix of text input **@textinput**
- [ ] 43c22d5 Fix QMLColor function and Rectangle color handling **@color**
- [ ] 91eb23c Now $delete also set's parent to undefined **@compdestr**
- [ ] 9dd3d75 Repeater bugfix **@repeater**
- [ ] 04afa40 Component.onDestruction **@compdestr**
- [ ] a743f6d Bugfix for Repeater **@repeater**
- [ ] 22ce160 Bugfix **@import**
- [ ] 281bf5b Qt.resolvedUrl function **@import**
- [ ] ac7ad8c Create engine.qmldirs and engine.qmldirsContent vars earlier **@import**
- [ ] 54de7de Allow to specify html tag name for component **@domtag**
- [ ] 48ace8c Import statemement test and small bugfix **@import**
- [ ] 4470475 addModulePath method and some fixes **@import**
- [ ] 409b7b5 Initial import implementation **@import**
- [ ] adea24c Keep info that import statement uses qualified dotted notation **@import**
- [ ] 5459537 Comment on how import statements work **@import**
- [ ] c65dae1 Bugfix in readQmlDir **@import**
- [ ] a339ee7 Root element resizing fixes **@sizing**
- [ ] 3deb3bb Allow Qt.createComponent to load from urls too **@import**
- [ ] 7ec1e3b Revert to old exception-style behavior in getUrlContents **@import**
- [ ] e89b83e Merge remote-tracking branch 'upstream/master'-
- [ ] 2131576 bugfix for engine::addImportPath **@import**
- [ ] d3b575c engine::setImportPathList method **@import**
- [ ] bc1ba46 Less error logging for Qt.createComponent **@import**
- [ ] 7475fce addImportPath etc **@import**
- [ ] 72b7834 getUrlContents changed to return "" on failed gets. **@import**

# Topics

**@sizing** - qml dom area size management / monitoring
**@domtag** - use something other than div for Item. **Important**.
**@repeater** - repeater bugfix. **Important**.
**@compdestr** - component onDistruction signal. **Important**.
**@color** - create color strings from numbers and arrays. **Important**.
**@textinput** - bugfix for TextInput. **Important**.
**@aliassignals** - aliases sends signals (e.g. aliasNameChanged). **Important**.
**@checkbox** - fix checkbox - change dom element `checked` state if qml element checked state changes. **Important**.
**@githubimport** - things related to loading qml files right from github. Experimental.
**@recursiveprops** - calculate properties recursively on init. Without that, all the binding process is incorrect. Maybe, this is already done? **Important**.
**@createqmlobject** - http://doc.qt.io/qt-5/qtqml-javascript-dynamicobjectcreation.html createQmlObject method. Must validate with current state of qmlweb. **Important**.
**@qobjectdelete** - must have delete method, according to Qml. **Important**.
**@video** - got from **@Plaristote**. Probably have some enhancements. 
**@binding** - binding bugfixes. Must validate with current state of qmlweb. **Important**.
**@debug** - things for easy debugging.
**@optimize** - optimizations.
**@fontbug** - font bugfix.
**@context** - bugfixes for context. Must validate with current state of qmlweb. **Important**.
**@outercontext** - outer context bugfix. **Important**.
**@signals** - signals bugfixes. **Important**.
**@bindingloop** - binding loop watchdog. **Important**.
**@textarea** - textarea fixes.
**@button** - button focus-unfocus signals.
**@upflow** - flow of identificators to upper contexts. Must stay as option. **Important**.
**@listmodel** - ListModel bugfix. **Important**.

**Comment by ChALkeR** [comment]
@pavelvasev I am afraid that you just accidently summoned a lot of unrelated people here. Could you edit this, please?

**Comment by pavelvasev** [comment]
Sorry I used wrong notation (e.g. @-symbol + topic name). I moved them to ** @-symbol + topic name**. Is that ok? ;-)

**Comment by ChALkeR** [comment]
@pavelvasev 
> I moved them to ** @ + topic name**. Is that ok?

No, they are still links. Could you remove the `@`, please? Something like `**[tag]**` would be ok.

**Comment by pavelvasev** [comment]
better?

**Comment by ChALkeR** [comment]
@pavelvasev Thanks!

**Comment by henrikrudstrom** [comment]
@pavelvasev made a test for the `upflow`, seems to pass, but please check if i test the right thing: 
commit is here: 7bacb63b1aefe75b76ec6414ec65c6076904e25e

**Comment by pavelvasev** [comment]
@henrikrudstrom test seems to be correct. Wow, it's great that it pass in current qmlweb!

**Comment by henrikrudstrom** [comment]
@pavelvasev can you describe of some test cases for some of the #bugsfixes? could help us figure out what to merge and make sure we dont introduce them again.

**Comment by pavelvasev** [comment]
There are a lot of bugfixes... And I agree, each of them better to accompany with a test case. However this is big work. @henrikrudstrom  may I ask you to be more specific now, which commits/topics looks most unclear at the moment?

**Comment by henrikrudstrom** [comment]
@pavelvasev dont sweat it, but if there are some that you can roughly remember. some of them would be caught by general render tests, but some tags that make me curious are: 
#context - bugfixes for context. Must validate with current state of qmlweb. Important.
#outercontext - outer context bugfix. Important.
#signals - signals bugfixes. Important.
#bindingloop - binding loop watchdog. Important.
#aliassignals - aliases sends signals (e.g. aliasNameChanged). Important.
#recursiveprops - calculate properties recursively on init. Without that, all the binding process is incorrect. Maybe, this is already done? Important.
#binding - binding bugfixes. Must validate with current state of qmlweb. Important.

**Comment by pavelvasev** [comment]
OK I will prepare tests for them first. BTW a lot of commits have large
comments in code, describing the problem they solve.
03.03.2016 1:24 пользователь "henrik" <notifications@github.com> написал:

**Comment by pavelvasev** [comment]
Hi @henrikrudstrom here are 2 test cases explaining #recursiveprops

https://github.com/qmlweb/qmlweb/compare/test-recursive-property-init

**Comment by henrikrudstrom** [comment]
great! I like tests that fail!

**Comment by henrikrudstrom** [comment]
you should add a pull request btw
add the test names to the list in failingTests.js to make everything green

**Comment by ChALkeR** [comment]
@pavelvasev Could you please take a look at the list at the very top of this issue and toggle the checkboxed that were merged in #68? I think I missed some of those.

**Comment by ChALkeR** [comment]
https://github.com/qmlweb/qmlweb/pull/68#issuecomment-192387387 — this is the list of commits in #68, for reference.

**Comment by pavelvasev** [comment]
@ChALkeR ok I will check and report tomorrow

**Comment by pavelvasev** [comment]
@ChALkeR I updated the lisf of commits at the top of the issue

**Comment by ChALkeR** [comment]
@pavelvasev Thanks!

**Comment by ChALkeR** [comment]
@henrikrudstrom It would be super helpful if you could file a PR with some more of those changes cherry-picked and rebased against master, like you did with #68 and #96. :-)

**Comment by ChALkeR** [comment]
I rejected 1ac2f5a because it's a copy of what we have now and 13af286 because it only brings in non-standard properties, those are not present in [the documentation](http://doc.qt.io/qt-5/qml-qtmultimedia-video-members.html). Not sure about `this.dom.style.pointerEvents = "all";`, but the css documentation says that's only valid for svg.

About non-standard extensions — we might want to support those in the future, but only by putting them into a separate add-on module, e.g. `QmlWeb.Dom`.

**Comment by ChALkeR** [comment]
@pavelvasev could you give an example which 816f334 does fix? I can't reproduce video content going out of the bounds with the current code.

**Comment by pavelvasev** [comment]
Hi @ChALkeR !

1. Sorry I don't remember what 816f334 fixes. 
2. pointer-events are not for svg only. The doc you had googled is bad-written.
3. I think video api of Qt/Qml is not excellent, that's why there are those props. However I agree with you that our enhancements might be present as separate layer.

**Comment by ChALkeR** [comment]
@pavelvasev Ok, thanks!

`pointer-events` are not for svg only, but `pointer-events: all` is. I updated my comment above, it looks like you have done that to negate `pointer-events: none` for controls to work. We don't need that there atm.

**Comment by ChALkeR** [comment]
@pavelvasev I think that 690c82b is not worh it at the current stage. Also, joining being faster than concatenating is correct for huge strings, I doubt that this would give us any visible improvement.

We need benchmarks first to do the optimization part non-blindfolded.

**Comment by ChALkeR** [comment]
**[recursiveprops]** are blocked by #195.

**Comment by ChALkeR** [comment]
It looks like dce7d96 changes in TextInput are not compatible with QML, there are no `onFocused`/onUnfocused` signals in QML for TextInput afaik. Rejecting that part.

**Comment by Plaristote** [comment]
About 871fbf0, I think we should re-consider it.
It postulates that we still have the extra executionContext parameter in the createComponent and createQmlObject methods, which we don't anymore.
Imho, there's nothing to keep in that commit anymore. If someone else can concur, we should reject it.

**Comment by ChALkeR** [comment]
@Plaristote @pavelvasev I am not sure what to do with 871fbf0, and I don't think I entirely understand what is going on there, so I leave that one in your hands =).

**Comment by ChALkeR** [comment]
That's the only commit here without a separate pull request, btw.

**Comment by ChALkeR** [comment]
871fbf0 is the last commit on all branches (excluding `kde-realinheritance`) that does not have a PR.

**Comment by ChALkeR** [comment]
Everything is merged here, except for this commit:
 * **[outercontext]** 561b202 propagate properties of component root objects to outer context

It breaks tests and is a one-liner, so I marked it as **Postponed** — we can deal with that one later, it shouldn't block 0.1.0 release and cleanups.

**Comment by ChALkeR** [comment]
Everything else was covered. Closing.
